### PR TITLE
[SPARK-26006][mllib] unpersist 'dataInternalRepr' in the PrefixSpan

### PR DESCRIPTION
**Pull request by shahidki31**
## What changes were proposed in this pull request?
Mllib's Prefixspan - run method - cached RDD stays in cache. After run is comlpeted , rdd remain in cache.
We need to unpersist the cached RDD after run method.

## How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @srowen Kindly review

**Comment by srowen** [comment]
I'm not sure about that, because the returned PrefixSpanModel has an RDD that depends on that RDD. We could cache the final RDD instead and materialize it; that could make more sense. 

In other places we have done such a thing only when the input is cached, in order to kind of follow the caller's lead, but there isn't a consistent standard for this.

I'd be OK improving this to persist the final RDD only, and then unpersist the intermediate one. That makes at least more sense. You can cache at the same storage level as the input (which might be NONE)

**Comment by shahidki31** [comment]
Thanks @srowen , I will update the PR

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala [hunk 1]
```diff
@@ -174,6 +174,10 @@ class PrefixSpan private (
     val freqSequences = results.map { case (seq: Array[Int], count: Long) =>
       new FreqSequence(toPublicRepr(seq), count)
     }
+    // Cache the final RDD to the same storage level as input
+    freqSequences.persist(data.getStorageLevel)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala]: The problem here is that it won't get persisted until something materializes it, and at that point its dependent RDD dataInternalRepr is already unpersisted.

I'd say that _if_ the input's storage level isn't NONE, then persist freqSequences at the same level and .count() it to materialize it. Then unpersist dataInternalRepr in all events.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by shahidki31** [comment]
Thank you @srowen

**Comment by idlevi** [comment]
Wouldn't the added count increase the execution time? did you do a benchmark?

**Comment by idlevi** [comment]
I am currently using it with spark 2.3 as 
 <dependency>
            <groupId>org.apache.spark</groupId>
            <artifactId>spark-mllib_2.11</artifactId>
        </dependency>
How can i get this fix?

**Comment by shahidki31** [comment]
@idlevi Actually, input and output of the prefix span are RDD. Earlier intermediate rdd was cached, now final rdd is cached, and materialized it. So, if you materialize the model, earlier it will compute from the intermediate level, now it directly get from the finalRdd.
 I ran all the UTs in the prefixSpanSuite, and there is hardly any time difference with/without the patch.

**Comment by shahidki31** [comment]
> I am currently using it with spark 2.3 as
> 
> org.apache.spark
> spark-mllib_2.11
> 
> How can i get this fix?

You can cherry pick the commit from the master branch.

**Pull request closed without merging**